### PR TITLE
Update ML-IAP for ACE support

### PR DESCRIPTION
**Pull request by jmgoff**
**Summary**

The ML-IAP package is updated to support ACE descriptors and models. Previously ML-IAP could not support ACE descriptor computes or ACE models (including linear, neural network, and more general non-linear models).  

Key enhancements and features are:

- LAMMPS-native support for arbitrary ACE pytorch models (including neural networks)
-- Users can run arbitrary ACE neural network poentials in LAMMPS, calculate energies, forces, and stresses
- Ability for users to separate linear model coefficients and clebsch-gordan coefficients (offers a more general format for ACE models developed from different software)
-- This may help serve as a template to simplify ACE inputto a single list of coefficients in the future
- Ability to use ACE descriptors in general mliappy/python routines (e.g. in structure generation algorithms, model training, and on-the-fly analysis)
- Various examples for ML-IAP + ACE are provided

**Related Issue(s)**

N/A

**Author(s)**

James Goff (Sandia National Laboratories) ( [goffjimmy1@gmail.com](mailto:goffjimmy1@gmail.com) )

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Will not break backward compatibility 

**Implementation Notes**

ACE support in mliap uses functions from the ML-PACE package, and updates were made to import new ACE descriptor evaluators implemented by the owners of the ML-PACE package. This largely required the development of a new mliap descriptor class in`src/ML-IAP/mliap_descriptor_ace.*` - and minor adjustments throughout the mliap package to accommodate the new class.

To validate the results for the mliap implementation of ACE descriptors and models:
The mliap ace compute reproduces descriptor values, descriptor gradients, and virial terms from`compute pace`. 
The energies, forces, and stresses were validated for mliap ACE models (including pytorch neural networks)  the `numdiff` fix.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by jmgoff** [comment]
Similar to the ML-IAP  python models requiring that PKG_PYTHON is installed, the ML-IAP ACE descriptors require that PKG_ML-PACE is also installed. This crashed some unit tests before because ML-PACE is not included by default.

Resolved by adding a new compilation flag: `-DMLIAP_ENABLE_ACE` that can only be turned on if `-DPKG_ML-PACE` is on.

To use ML-IAP & ACE functionality, compile lammps with the following dflags (in addition to the dflags required for Python and ML-IAP:
`-DPKG_ML-PACE=yes`
`-DMLIAP_ENABLE_ACE=yes`

**Review by akohlmey — CHANGES_REQUESTED** [review]
This will need updates to the documentation.
Also, I only see support for the CMake build system. What about the traditional build with GNU make?

**Comment by akohlmey** [comment]
Since both ML-IAP and ML-PACE have KOKKOS versions and also mliappy has KOKKOS support, does this PR need some KOKKOS updates?

**Comment by athomps** [comment]
> Since both ML-IAP and ML-PACE have KOKKOS versions and also mliappy has KOKKOS support, does this PR need some KOKKOS updates?

The same functionality already exists for SNAP descriptors, but these have not yet been Kokkosized in MLIAP.  I think this will happen soon, but until then, it is better not to attempt it for ACE descriptors, which are quite a bit more complicated than the SNAP descriptors

File: cmake/Modules/Packages/ML-IAP.cmake [hunk 1]
```diff
@@ -10,6 +10,15 @@ endif()
 
 option(MLIAP_ENABLE_PYTHON "Build ML-IAP package with Python support" ${MLIAP_ENABLE_PYTHON_DEFAULT})
 
+# if ML-PACE package is included we may also include ML-PACE support in ML-IAP
+set(MLIAP_ENABLE_ACE_DEFAULT OFF)
+if(PKG_ML-PACE)
+  set(MLIAP_ENABLE_PYTHON_DEFAULT ON)
+endif()
+
+option(MLIAP_ENABLE_PYTHON "Build ML-IAP package with Python support" ${MLIAP_ENABLE_PYTHON_DEFAULT})
```
Comment (akohlmey) [inline cmake/Modules/Packages/ML-IAP.cmake]: This is replicated from line 11

File: src/ML-IAP/mliap_descriptor_ace.cpp [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+  }
+  ACECTildeBasisSet *basis_set;
+  ACECTildeEvaluator *ace;
+};
+}
+
+
+using namespace LAMMPS_NS;
+
+#define MAXLINE 1024
```
Comment (akohlmey) [inline src/ML-IAP/mliap_descriptor_ace.cpp]: These two defines are not used anywhere.

**Comment by akohlmey** [comment]
> but until then, it is better not to attempt it for ACE descriptors

Then there should be some tests implemented that prevent using this with KOKKOS or someone needs to explain how this cannot produce problems with the current code if somebody tries this.

File: src/ML-IAP/mliap_descriptor_ace.cpp [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+#include "memory.h"
+#include "mliap_data.h"
+#include "neigh_list.h"
+#include "pair_mliap.h"
+#include "tokenizer.h"
+
+#include <cmath>
+#include <cstring>
+#include <typeinfo>
+//#include <map>
```
Comment (akohlmey) [inline src/ML-IAP/mliap_descriptor_ace.cpp]: Please remove commented out code like this.

File: src/ML-IAP/mliap_descriptor_ace.cpp [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+#include "error.h"
+#include "memory.h"
+#include "mliap_data.h"
+#include "neigh_list.h"
+#include "pair_mliap.h"
+#include "tokenizer.h"
+
+#include <cmath>
+#include <cstring>
+#include <typeinfo>
```
Comment (akohlmey) [inline src/ML-IAP/mliap_descriptor_ace.cpp]: Is this include file needed?

**Comment by stanmoore1** [comment]
I am OK with Aidan's suggestion of waiting on adding Kokkos support for ACE/SNAP. So I think this PR can be merged after addressing Axel's feedback.

**Comment by akohlmey** [comment]
@jmgoff FYI, I've merged this branch with upstream and resolved a merge conflict due to recently added functionality.

**Review by athomps — APPROVED** [review]
I approve

**Review by stanmoore1 — CHANGES_REQUESTED** [review]
Need to error out if a user tries to use the Kokkos version of ML-IAP with ACE or SNAP descriptors, something like this:

```
if (lmp->kokkos) error->all(FLERR,"Cannot (yet) use KOKKOS package with SNAP descriptors");
```

**Comment by jmgoff** [comment]
Thank you all for the comments and reviews. The most recent pushes should address the remaining comments and issues. To summarize:
For Axel's previous comments:
-Documentation has been updated and added for ace in MLIAP
-Cleaned up leftover comments, defines, and whitespace in docs
-Support for make added. Updated ML-IAP make file setup from `ML-IAP` to allow for optional ace descriptor use (requiring that ML-PACE is also built) 
For Aidan's comments:
-Standard g++ (1 core) and (4 core) outputs have been added for the new MLIAP/ace examples
To address comments from Axel, Stan, and Aidan on the use of kokkos:
-Following Stan's comment this morning and previous discussion, checks have been added to `pair_mliap.cpp` and `compute_mliap.cpp` to throw runtime errors when users attempt to use `mliap/kk` with SNAP or ACE in mliap computes or pair styles.

**Comment by akohlmey** [comment]
@jmgoff both CMake and GNU make support were incorrect or incomplete. I have tried to fix it.

There were still some style issues that I resolved and I also applied clang-format to make the programming style consistent. I also reordered the initializer list of the MLIAPData class to avoid compiler warnings.

Since CPU snap descriptors are no longer allowed for KOKKOS, they corresponding unit tests need to be skipped.

**Review by akohlmey — APPROVED** [review]

**Comment by jmgoff** [comment]
> @jmgoff both CMake and GNU make support were incorrect or incomplete. I have tried to fix it.
> 
> There were still some style issues that I resolved and I also applied clang-format to make the programming style consistent. I also reordered the initializer list of the MLIAPData class to avoid compiler warnings.
> 
> Since CPU snap descriptors are no longer allowed for KOKKOS, they corresponding unit tests need to be skipped.

Hi Axel, thanks for applying the extra corrections for the CMake and gmake scripts along with the styling issues. I'll keep an eye out for similar things in the future.

As long as you and @stanmoore1 are okay with disabling mliap/kk + CPU snap for now, this sounds good to me.

**Comment by stanmoore1** [comment]
Does it need to be disabled for Kokkos? i.e. does it give the wrong answer, or is it just slow but still works correctly? I thought it was giving the wrong answer or couldn't run, but if it is just slow then maybe we just keep it?

**Comment by jmgoff** [comment]
That is a good point. SNAP is only crashes for the mliappy model, not for the linear, quadratic, or NN models. I'll add in errors for the mliappy model specifically to make sure that we don't have to turn off unit tests that work.

**Comment by jmgoff** [comment]
Adjusted it to only have ACE + mliap kokkos disabled and turned unit tests back on for SNAP + mliap kokkos. The SNAP mliappy model example works as well with the SNAP CPU descriptors.

**Comment by akohlmey** [comment]
@stanmoore1 this is now waiting on you

File: doc/src/pair_mliap.rst [hunk 5]
```diff
@@ -169,6 +198,24 @@ contains a few more arguments (e.g., *nmax* and *alpha*). The preparation
 of SO3 descriptor and model files can be done with the
 `Pyxtal_FF <https://github.com/qzhu2017/PyXtal_FF>`_ package.
 
+The ACE descriptor file differs from the SNAP and SO3 files. It more
+closely resembles the potential file format for linear or square-root
+embedding ACE potentials used in the :doc:`pair_style pace <pair_pace>`.
+As noted above, the key difference is that the Clebsch-Gordan coefficients
+in the descriptor file with *mliap descriptor ace* are *NOT* multiplied
+multiplied by linear or square root embedding terms. In other words,
```
Comment (stanmoore1) [inline doc/src/pair_mliap.rst]: Typo: "multiplied multiplied" (occurs twice)

File: doc/src/pair_mliap.rst [hunk 6]
```diff
@@ -72,6 +73,24 @@ the smooth SO(3) power spectrum with the explicit inclusion of a radial
...
+atomic cluster expansion descriptors (ACE) from :ref:`(Drautz) <Drautz19>` ,
+that include a radial basis, an angular basis, and bases for other variables
+(such as chemical species) if relevant. In descriptor style *ace*, the *ace*
+descriptors may be defined up to an arbitrary body order. This descriptor
+style is the same as that used in :doc:`pair_style pace <pair_pace>` and
+:doc:`compute pace <compute_pace>`. The available models with *ace* in
+ML-IAP are *linear* and *mliappy*. The *ace* descriptors and models require
+building LAMMPS with the ML-PACE package (see below). The *mliappy* model
+style may be used with *ace* descriptors, but it requires that LAMMPS is
+also built with the PYTHON package. As with other, the *mliappy* model style
```
Comment (stanmoore1) [inline doc/src/pair_mliap.rst]: "As with other," ... not sure what this means or what "other" refers to?

**Review by stanmoore1 — APPROVED** [review]
A few minor doc issues, otherwise looks good.
